src/Common/src/System/Collections/Generic/EnumerableHelpers.cs

```diff
@@ -42,53 +40,69 @@ internal static T[] ToArray<T>(IEnumerable<T> source, out int length)
-                        if (newLength < count + 1)
-                        {
-                            newLength = count + 1;
+                            if (count == arr.Length)
+                            {
+                                // MaxArrayLength is defined in Array.MaxArrayLength and in gchelpers in CoreCLR.
+                                // It represents the maximum number of elements that can be in an array where
+                                // the size of the element is greater than one byte; a separate, slightly larger constant,
+                                // is used when the size of the element is one.
+                                const int MaxArrayLength = 0x7FEFFFFF;
+        
+                                // We attempt to 
+                                // double the size of the array. Before doubling overflows, it will exceed MaxArrayLength, so it
+                                // is constrained to that size.
+                                // Because a slightly larger constant is used when T is one byte in size, we
+                                // could then end up in a situation where arr.Length is MaxArrayLength or slightly larger, and an
+                                // overflow does happen.
+                                // In this case we increase the array by 2383 each resize as the smallest factor above square root
+                                // of the difference between the maximum one-byte array size (0x7FFFFFC7) and MaxArrayLength as a
+                                // compromise between going straight to 0x7FFFFFC7 (fewer allocations) or adding only one element
+                                // at a time (smaller allocations).
+                                int newLength = count << 1;
+                                if ((uint)newLength > (uint)MaxArrayLength)
+                                {
+                                    if (count < MaxArrayLength)
+                                    {
+                                        newLength = MaxArrayLength;
+                                    }
+                                    else if ((newLength = unchecked(count + 2383)) < 0)
+                                    {
+                                        newLength = int.MaxValue; // Just let the attempt to create an array this size fail.
+                                    }
+                                }
+        
+                                Array.Resize(ref arr, newLength);
+                            }
+
+                            arr[count++] = en.Current;
                         }
 
-                        Array.Resize(ref arr, newLength);
+                        length = count;
+                        return arr;
                     }
-                    arr[count++] = item;
                 }
             }
 
-            length = count;
-            return arr;
+            length = 0;
+            return Array.Empty<T>();
```

Comment:
No. `EnumerableHelper` is unchanged in this regard (the only observable change is that it might have a larger excess with one-byte-element arrays larger than 0x7FEFFFFF) and while using `EnumerableHelper` to build `Buffer` means it would internally have a reference to `Array.Empty` where before it would have had `null` in actually using the array it still pass through [https://github.com/dotnet/corefx/pull/3021/files#diff-95a14d3b4ae14339b623dd95c7537fd0R3398](the check):

    if (count == 0) return new TElement[0]; // consider replacing with Array.Empty<TElement>()

If the possibility in the comment is ever decided as the way to go, then this now means just removing that line, but in the meantime a fresh zero-length array is still created each time.